[BOM-804] feat: 챌린지 신청 및 취소 API 구현

kysub99 opened the pull request:
[변경범위](https://github.com/woowacourse-teams/2025-bom-bom/pull/598/files/99b5ea77efbab3624230c93fc44fd515dee4d2fe..ec86a2cb98285f71082e87ed86bcfb42b773cefd)
## 📌 What
<!-- 해결하고자 한 문제를 간결하고 명확하게 서술해주세요. -->
<!-- ex) 로그인 실패 시 에러 메시지가 출력되지 않던 문제 -->
챌린지 신청 / 취소 API 추가
- POST /challenges/{id}/application
- DELETE /challenges/{id}/application

## ❓ Why
<!-- 이 문제를 왜 해결해야 했는지, 어떤 맥락에서 발생했는지 작성해주세요. -->
<!-- ex) 사용자 피드백으로 오류 발생 시 원인 파악이 어렵다는 의견이 있었음 -->
챌린지 신청을 위해

## 🔧 How
<!-- 어떤 방식으로 해결했는지 기술적 접근 방법을 설명해주세요. -->
<!-- 주요 변경 사항, 선택한 로직이나 방법에 대한 이유가 있으면 함께 작성해주세요. -->
Eligibility 판단 로직 정리
- validateEligibility 메서드는 상태 판단만 담당하고 EligibilityReason을 반환
- 조회 API(checkEligibility)에서는 상태 응답으로 사용
- 신청 API(applyChallenge)에서는 상태에 따라 예외를 생성

챌린지 목록 조회 성능 개선
- 기존: member 기준 전체 참여 조회
- 변경: challengeId IN 기반 조회로 범위 제한

## 👀 Review Point (Optional)
<!-- 리뷰어에게 중점적으로 봐주었으면 하는 부분을 작성해주세요. -->
<!-- ex) 예외 처리 방식이 적절한지 확인 부탁드립니다. -->
잘못된 로직 있는지 위주로 부탁드립니다
(Eligibility 판단 로직 부분을 Policy 같이 다른 객체로 분리해도 괜찮을 것 같다고 생각은 했는데, 시간 문제로 일단 서비스 내부에 다 구현 했습니다.)

kysub99 committed:
feat: 챌린치 신청 및 취소 기능 구현

kysub99 committed:
feat: 챌린치 신청 및 취소 기능 구현

kysub99 committed:
test: 챌린치 신청 및 취소 기능 서비스 테스트

kysub99 committed:
test: 챌린치 신청 및 취소 기능 서비스 테스트

kysub99 committed:
test: 챌린치 신청 및 취소 파라미터 검증 테스트

kysub99 committed:
test: 챌린치 신청 및 취소 파라미터 검증 테스트

kysub99 committed:
fix: 챌린지 신청 및 취소 상태 코드 변경

kysub99 committed:
fix: 챌린지 신청 및 취소 상태 코드 변경

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/service/ChallengeService.java:
```diff
@@ -88,21 +90,99 @@ public ChallengeEligibilityResponse checkEligibility(Long challengeId, Member me
+        }
+
+        ChallengeParticipant participant = challengeParticipantRepository
+                .findByChallengeIdAndMemberId(challengeId, member.getId())
```
p1

여기 개행이 잘못됐어요!

kysub99 replied:
https://github.com/woowacourse-teams/2025-bom-bom/pull/598/commits/626b345d3fc40e4070ee9d1eda263eb7ea2eec12

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/service/ChallengeService.java:
```diff
@@ -88,21 +90,99 @@ public ChallengeEligibilityResponse checkEligibility(Long challengeId, Member me
+        challengeParticipantRepository.save(participant);
+    }
+
+    private RuntimeException createApplyException(Long challengeId, Member member, EligibilityReason reason) {
```
p1

private 메서드 아래로 내려주세요!

kysub99 replied:
https://github.com/woowacourse-teams/2025-bom-bom/pull/598/commits/626b345d3fc40e4070ee9d1eda263eb7ea2eec12

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/service/ChallengeService.java:
```diff
@@ -88,21 +90,99 @@ public ChallengeEligibilityResponse checkEligibility(Long challengeId, Member me
+                    .addContext(ErrorContextKeys.ENTITY_TYPE, "challengeParticipant")
+                    .addContext(ErrorContextKeys.MEMBER_ID, member.getId())
+                    .addContext("challengeId", challengeId);
+        }
```
p2

멱등성을 보장하기 위해 이미 가입된 것은 예외를 터트리지 않고 성공 응답을 주는게 어떠신가여

kysub99 replied:
이미 신청 됐을 때는 debug 수준으로 로그만 남기고 정상 응답 주도록 수정했습니다

https://github.com/woowacourse-teams/2025-bom-bom/pull/598/commits/ec86a2cb98285f71082e87ed86bcfb42b773cefd

Choidongjun0830 commented on backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/service/ChallengeService.java:
```diff
@@ -88,21 +90,99 @@ public ChallengeEligibilityResponse checkEligibility(Long challengeId, Member me
+                    .addContext("challengeId", challengeId);
+        }
+        if (reason == EligibilityReason.NOT_SUBSCRIBED) {
+            return new CIllegalArgumentException(ErrorDetail.INVALID_INPUT_VALUE)
```
구독 안되어 있는건 왜 `INVALID_INPUT_VALUE`인가여?

kysub99 replied:
기존에 있던 ErrorDetail 최대한 쓰려고 했는데, 너무 모호한거 같아서 사전조건이 만족되지 않았음을 나타내는 `PRECONDITION_FAILED`을 추가했습니다.

https://github.com/woowacourse-teams/2025-bom-bom/pull/598/commits/dff87c24a478c765a512151660cbf7e8080ba6a2

Choidongjun0830 reviewed (changes requested):
수고하셨습니다! 4개 리뷰 코멘트 남겨뒀습니다!

kysub99 committed:
style: 코드 컨벤션

rladmstn committed:
[BOM-782] feat: 챌린지 한 줄 코멘트 목록 조회 API 구현 (#594)

* feat: 챌린지 한 줄 코멘트 목록 조회 API 구현

* chore: 챌린지 한 줄 코멘트 테이블에 quotation 필드 추가

* test: 챌린지 한 줄 코멘트 목록 조회 테스트 작성

* Revert "[BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)"

This reverts commit 0e0ccdd7be5da65a5d9baa6be751ef3bdcd0a4c2.

* Revert "Revert "[BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)""

This reverts commit 3fe1fd134904b5970586348f38a1b9961bc5afc7.

* feat: 코멘트 목록 조회 시 DTO에 해당 뉴스레터 구독 여부 추가

* feat: 챌린지 코멘트 목록 조회 시 날짜 param @ModelAttribute로 통합

* test: 테스트 관련 클래스 import 추가

* test: 컨트롤러 테스트 시 응답코드만 검증하도록 수정

* test: ChallengeComment TestFixture 추가

* feat: ChallengeComment에 뉴스레터 id, 아티클 제목 저장하도록 수정

* [BOM-780] refactor: 지난 뉴스레터 관리 Role 기반 관리로 수정 (#590)

* refactor: 지난 아티클 아카이빙을 Role ID 기반으로 동작하도록 수정

* test: 지난 아티클 아카이빙을 Role ID 기반 테스트 작성

* feat: RoleRepository 추가

* refactor: 아카이빙 관리자 id 조회 로직 수정

* style: 개행 수정

* [BOM-784] feat: 챌린지 상세 조회 API 구현 (#593)

* feat: 챌린지 상세 조회 API 구현

* feat: PathVariable 검증 예외 처리 추가

* feat: 챌린지 상세 조회 컨트롤러 단위 테스트 작성

* feat: 챌린지 상세 조회 서비스 통합 테스트 작성

* fix: API 엔드포인트 수정

* feat: 응답 DTO에 제약 추가

* style: 응답 DTO 네이밍 수정

* refactor: 에러 핸들링 메서드 분리 및 에러 메시지 수정

* style: 테스트 클래스명 수정

* test: API 엔드포인트 수정에 따른 테스트 수정

* chore: 스웨거 수정

* fix: 최소 수료 조건 로직 수정

* fix: 요청 파라미터 검증 메서드 중복 제거

* fix: 탈퇴한 회원에 대해 닉네임 null로 반환

* fix: 챌린지 코멘트 응답에 notnull 제약 제거

* [BE][BOM-786] feat: 챌린지 전체 조회 API (#592)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* [BOM-803] feat: 챌린지 신청 가능 여부 조회 api (#597)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

* feat: 챌린지 신청 권한 확인 API

* test: Challenge 서비스 테스트

* test: Challenge 컨트롤러 단위 테스트

* refactor: repository에서 챌린지 신청 여부 확인

* [BE][BOM-786] feat: 챌린지 전체 조회 API (#592)

* feat: 챌린지 도메인 및 전체 조회 API 개발

* test: 챌린지 도메인 테스트 작성

* chore: 미사용 import 제거

* [BOM-783] feat: 챌린지  관련 도메인 생성 (#591)

* feat: 챌린지 관련 도메인 생성

* fix: unique 제약 수정

* [BE][BOM-654] feat: 아티클 자동 삭제 스케줄러 추가 (#588)

* feat: 아티클 자동삭제 스케줄러 추가

* feat: 아티클 삭제 시 북마크 제외 및 관리자 제한 분기

* test: 아티클 삭제 스케줄러 테스트 작성

* chore: 버저닝 오타 수정

* fix: 아티클 자동 삭제 시 벌크 연산으로 변경

* style: 불필요한 세미콜론 제거

* fix: 아티클 정리 시 아티클 남겨둘 최소 개수 추가

* feat: 아티클 삭제 시 최소 개수 검증 추가

---------

Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* refactor: ChallengeGrade enum 값명 변경

* style: 코드 컨벤션

* refactor: LocalDate.now() 호출을 필요 시점에 직접 호출하도록 변경

* refactor: 불필요한 else 제거

* refactor: toList 로 변경

* refactor: ChallengeParticipant 파라미터 타입 변경

* refactor: ChallengeNewsletter 조회 로직을 Repository로 이동

* chore: 충돌 해결

* feat: 챌린지 조회시 기수 포함

* feat: 챌린지 뉴스레터 목록 정렬

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* feat: 챌린지 신청 권한 확인 API

* test: Challenge 서비스 테스트

* test: Challenge 컨트롤러 단위 테스트

* refactor: repository에서 챌린지 신청 여부 확인

* refactor: 챌린지 뉴스레터 구독 확인 쿼리 단순화

* style: 컨벤션

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: Kim Eunsu <88280787+rladmstn@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

* feat: 챌린지 한 줄 코멘트 목록 조회 API 구현

# Conflicts:
#	backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/repository/ChallengeParticipantRepository.java

* test: 챌린지 참여자 TestFixture 추가

* fix: conflict 해결

---------

Co-authored-by: ChoiDongjun <adong0808@naver.com>
Co-authored-by: geonwoo kim <115832836+kysub99@users.noreply.github.com>
Co-authored-by: Cheol Won <76567238+Ryan-Dia@users.noreply.github.com>

kysub99 committed:
refactor: 챌린지 뉴스레터 미구독에 대한 ErrorDetail 수정

kysub99 committed:
fix: 챌린지 재신청시 정상 응답

kysub99 committed:
Merge remote-tracking branch 'origin/BOM-804-챌린지-신청-신청-취소-api' into BOM-804-챌린지-신청-신청-취소-api

# Conflicts:
#	backend/bom-bom-server/src/main/java/me/bombom/api/v1/challenge/service/ChallengeService.java
#	backend/bom-bom-server/src/test/java/me/bombom/api/v1/challenge/service/ChallengeServiceTest.java

kysub99 committed:
chore: conflict 해결

rladmstn reviewed (approved):
수고하셨습니다!
다른 리뷰 코멘트들 잘 반영해주신 것 같아서 로직 괜찮아보입니다 👍

Choidongjun0830 reviewed (approved):
귯 수고하셨습니다
